geozero/src/wkt/wkt_writer.rs

```diff
@@ -178,7 +178,16 @@ impl<W: Write> GeomProcessor for WktWriter<'_, W> {
 
 impl<W: Write> PropertyProcessor for WktWriter<'_, W> {}
 
-impl<W: Write> FeatureProcessor for WktWriter<'_, W> {}
+impl<W: Write> FeatureProcessor for WktWriter<'_, W> {
+    // Feature Collections may contain multiple geometries, so we need to wrap
+    // them in a GeometryCollection to ensure we output valid geometry.
+    fn dataset_begin(&mut self, _name: Option<&str>) -> Result<()> {
+        self.geometrycollection_begin(0, 0)
```

Comment:
I think here's a confiusion between FeatureCollections and GeometryCollections. The latter is a special geometry type for _one_ feature (https://datatracker.ietf.org/doc/html/rfc7946#appendix-A.7).

Reply:
I'll respond just here - since I think all three of your comments are essentially about the same issue.

> I think here's a confusion between FeatureCollections and GeometryCollections

This was intentional - I am proposing that we explicitly interpret a feature collection as a geometry collection in WKT.

You're right that in geojson `GeometryCollection` and `FeatureCollection` are different things. But WKT, being concerned only with Geometry, has no concept of `Feature` or `FeatureCollection`. 

So then, if we get a FeatureCollection from geojson and want to convert it to WKT, how should we do that?

Our current approach is to output a series of comma separated WKT geometries. Unfortunately, this output cannot be parsed by postgis, gdal, jts, nor georust::wkt (thus not by geozero itself!).

A corollary exists with the [geo-types geometry writer](https://github.com/georust/geozero/blob/master/geozero/src/geo_types/geo_types_writer.rs#L36). Similar to WKT, geo-types is strictly Geometry - no Feature/FeatureCollections. And in the same way, a geojson FeatureCollection with multiple features will be output as a top level `geo_types::GeometryCollection`.

Do you see a practical downside with the approach I'm proposing here?

Reply:
I'll respond just here - since I think all three of your comments are essentially about the same issue.

> I think here's a confusion between FeatureCollections and GeometryCollections

This was intentional - I am proposing that we explicitly interpret a feature collection as a geometry collection in WKT.

You're right that in geojson `GeometryCollection` and `FeatureCollection` are different things. But WKT, being concerned only with Geometry, has no concept of `Feature` or `FeatureCollection`. 

So then, if we get a FeatureCollection from geojson and want to convert it to WKT, how should we do that?

Our current approach is to output a series of comma separated WKT geometries. Unfortunately, this output cannot be parsed by postgis, gdal, jts, nor georust::wkt (thus not by geozero itself!).

A corollary exists with the [geo-types geometry writer](https://github.com/georust/geozero/blob/master/geozero/src/geo_types/geo_types_writer.rs#L36). Similar to WKT, geo-types is strictly Geometry - no Feature/FeatureCollections. And in the same way, a geojson FeatureCollection with multiple features will be output as a top level `geo_types::GeometryCollection`.

Do you see a practical downside with the approach I'm proposing here?

Reply:
You're right that this affects other output formats. My goal is to make sure all the "geometry only" formats — those which don't allow feature data: geo-types, WKT, WKB — behave sanely.

For example, he's something I consider a problem with the current `to_wkb()` implementation:

As input, consider this geojson feature collection with two Point features.

```
let geojson = r#"{        
    "type": "FeatureCollection",                
    "features": [                
        {                
            "type": "Feature",                
            "properties": {                
                "population": 100                
            },                
            "geometry": {                
                "type": "Point",                
                "coordinates": [10.0, -20.0]                
            }                
        },                
        {                
            "type": "Feature",                
            "properties": {                
                "population": 100                
            },                
            "geometry": {                
                "type": "Point",                
                "coordinates": [30.0, -40.0]                
            }                
        },                
    ]                
} "#;
```

Then this contrived use case:

```
let sql = format!("SELECT ST_Centroid({}::geometry)", geojson.to_wkb());
```

Does that seem like a reasonable use case? What answer would you expect?


Currently it's:
``` 
$ SELECT ST_AsText(wkb_geometry) FROM( SELECT
    ST_Centroid('0101000000000000000000244000000000000034c001010000000000000000003e4000000000000044c0'::geometry) 
    AS wkb_geometry
) AS f;

   st_astext   
---------------
 POINT(10 -20)    <---- 🚨🚨🚨This is not what I'd expect!
```


Taking a closer look at the value of `geojson.to_wkb():
```
0101000000000000000000244000000000000034c001010000000000000000003e4000000000000044c0    
```

To break that down, the first part is:
```
// WKB of POINT(10.0, -20.0)
0101000000000000000000244000000000000034c0                                              
```

concatenated directly to:
```
// WKB of POINT(30.0, -40.0)
01010000000000000000003e4000000000000044c0    
```

I don't know if it's valid WKB to simply concatenate multiple WKB's together, but for better or worse, postgis actually seems to parse it:
```
$ SELECT ST_AsText(wkb_geometry)  FROM ( SELECT
    '0101000000000000000000244000000000000034c001010000000000000000003e4000000000000044c0'::geometry 
    AS wkb_geometry
) AS f;
   st_astext   
---------------
 POINT(10 -20)
```

The second point has apparently just been ignored!

With the changes proposed in this PR, just like with the WKT example, the output of FeatureCollection -> WKB has changed to be wrapped in a GeometryCollection.

So now its:
```
$ SELECT ST_AsText(wkb_geometry) FROM (  SELECT
    '0107000000020000000101000000000000000000244000000000000034C001010000000000000000003E4000000000000044C0'::geometry 
    AS wkb_geometry
) AS f;
                    st_astext                    
-------------------------------------------------
 GEOMETRYCOLLECTION(POINT(10 -20),POINT(30 -40))
```                                                                                        

I think this is much more likely to behave how the user expects. The centroid selected from the example will be the centroid of all the geometries from all the features.

```
$ SELECT ST_AsText(wkb_geometry)  FROM ( SELECT
    ST_Centroid('0107000000020000000101000000000000000000244000000000000034C001010000000000000000003E4000000000000044C0'::geometry)
    AS wkb_geometry
) AS f;

   st_astext   
---------------
 POINT(20 -30)
```

Do you agree that the current situation is problematic, or is my use case overly contrived?

Do you think there's a different solution?

FYI I pushed up another commit with some tests that show the new wkb behavior.

Reply:
If this approach generally makes sense to you, I can add some more thorough testing to the other output formats.

Reply:
Thanks for the explanation and the good example. The main problem is IMO, that we currently do not distinct between multi-feature (dataset) capable formats like GeoJSON and single feature formats like WKB. It's a good moment to find a proper solution for that, since I'm replicating the `GeozeroDatasource` and `FeatureProcessor` traits based on the new `GeomEventProcessor` in the coming days. What you're describing is collecting geometries from multiple features into a single feature GeometryCollection. I also think that's a valid use case, but maybe it should be made explict for the user. My current idea for the event API is that we'll have in addtion to `to_wkb()` an extended method `to_wkb_with(v: dyn GeomVisitor)`, which will allow to pass a processing chain e.g. including a reprojection. Maybe a `GeometryCollector` could also be a part of a `GeomVisitor` chain, or it could be a visitor flag, or a special method like `collect_to_wkb()`.

Reply:
> collecting geometries from multiple features into a single feature GeometryCollection. I also think that's a valid use case, but maybe it should be made explict for the user.

It seems like you have some reservations about this - so I want to make sure we're on the same page.

Setting aside implementation details for now, and just focusing on behavior: I think it's a good expectation that all format conversions (e.g. `to_wkb()`) should produce valid output by default without needing to pass in any kind of special options. Why would someone want invalid wkb? Do you agree with that part?

Test cases and examples can be helpful. Can you help think of some code that behaves surprisingly with the proposed behavior - I think that'd help me understand your reservations about making this default behavior. And maybe then I could be able to address them.

Reply:
Are there any questions I can answer or changes I can make to move this forward?

(Also, I just now rebased against latest.)